mne/io/base.py

```diff
@@ -2562,98 +2536,49 @@ def set_annotations(self, annotations):
 
 ###############################################################################
 # Writing
-def _write_raw(
-    fname,
-    raw,
-    info,
-    picks,
-    fmt,
-    data_type,
-    reset_range,
-    start,
-    stop,
-    buffer_size,
-    projector,
-    drop_small_buffer,
-    split_size,
-    split_naming,
-    part_idx,
-    prev_fname,
-    overwrite,
-):
-    """Write raw file with splitting."""
-    # we've done something wrong if we hit this
-    n_times_max = len(raw.times)
-    if start >= stop or stop > n_times_max:
-        raise RuntimeError(
-            "Cannot write raw file with no data: %s -> %s "
-            "(max: %s) requested" % (start, stop, n_times_max)
-        )
 
-    # Expand `~` if present
-    fname = str(_check_fname(fname=fname, overwrite=overwrite))
+# Assume we never hit more than 100 splits, like for epochs
+MAX_N_SPLITS = 100
+
 
-    base, ext = op.splitext(fname)
-    if part_idx > 0:
-        if split_naming == "neuromag":
-            # insert index in filename
-            use_fname = "%s-%d%s" % (base, part_idx, ext)
+def _write_raw(raw_fid_writer, fpath, split_naming, overwrite):
+    """Write raw file with splitting."""
+    dir_path = fpath.parent
+    # We have to create one extra filename here to make the for loop below happy,
+    # but it will raise an error if it actually gets used
+    split_fnames = _make_split_fnames(
+        fpath.name, n_splits=MAX_N_SPLITS + 1, split_naming=split_naming
+    )
+    is_next_split, prev_fname = True, None
+    for part_idx in range(0, MAX_N_SPLITS):
+        if not is_next_split:
+            break
+        bids_special_behavior = part_idx == 0 and split_naming == "bids"
+        if bids_special_behavior:
+            reserved_fname = dir_path / split_fnames[0]
+            logger.info(f"Reserving possible split file {reserved_fname.name}")
+            _check_fname(reserved_fname, overwrite)
+            reserved_ctx = _ReservedFilename(reserved_fname)
```

Comment:
@dmalt in addition to changing from `while` to `for`, I also DRY'ed the code a bit. The previous version had a first `if` clause that had a lot of the same code in it that the `while` loop had. By embedding the logic for `bids_special_naming` *inside* the loop, we can remove that redundancy. Now I think it's clearer when something special has to happen. WDYT?

Reply:
Honestly,  I'm not the biggest fan of the DRY principle. I feel like it definitely must be applied when there are >= 3 repetitions, but when something is repeated 2 times, it's sometimes better to resist the urge and just repeat yourself. For me here DRY doesn't improve readability because the control flow is more complex. Here I guess it's a matter of taste, so I'll be happy with whichever option you choose. Just saying it for the sake of discussion.

Reply:
> DRY doesn't improve readability because the control flow gets more complex.

It is true that there are now two `if` statements instead of one. But in the old code the *conditional variable values* that end up needing to be accounted for (in the mental model and executed in code) were more plentiful: `part_idx`, `prev_fname`, and `is_next_split` could all be modified inside that first conditional, so all of their values were essentially conditional on (what is now known as) `bids_special_behavior`, even though setting those values was accomplished with a single `if` statement.

In the new version of the code that uses a unified loop, this isn't the case: `part_idx` always starts at zero and increases in one place (`range`); `prev_fname` starts as `None` and gets set to the `current_fname` or whatever at the end of each iteration of the loop; `is_next_split` only gets overwritten one place (inside the loop) instead of two. The only special/conditional thing that happens is that for BIDS there might be a renaming, which leads to two "if" statements, but to me the variable behavior and otherwise equivalence of the behaviors under the two naming schemes is more readily apparent. But I guess this could be based on a different emphasis / mental model for how similar these two schemes are similar rather than how they differ...

Reply:
I was thinking of  `prev_fname` more as of explanatory variable. It could be plugged directly into `_write_raw_data()`, e.g. as keyword arguments to be more explicit. I see the point with regards to `part_idx` and `is_next_split` but I feel like with duplication you just read the code from top to bottom and in dried version there's more mental context to keep. Especially with the `with start..., reserved_ctx:` line for which you need to go back and to reassure yourself, that it's triggered only for the 0-th iteration and you're not actually reserving the filename every time.  For me the file reserving behaviour is the hardest part to understand and an extra conditional on it doesn't help.

> But I guess this could be based on a different emphasis / mental model for how similar these two naming schemes are similar rather than how they differ...

I agree. And readability has a lot of subjectiveness to it in general. I've just started exploring this 'DRY being not always good' idea and I was curious where you stand on it.

Reply:
>  'DRY being not always good' idea and I was curious where you stand on it.

Agreed it's not always good, just one of the many things to weigh when deciding how to implement something!

Reply:
BTW, both cyclomatic and cognitive complexity give a little lower scores for the duplicated version: 7 vs 8 and 9 vs 11 correspondingly.